trie-db/src/iterator.rs

```diff
@@ -0,0 +1,647 @@
+                        self.db.get_raw_or_lookup(item, prefix.left())?
+                    },
+                    Node::Branch(nodes, _) => {
+                        if partial.is_empty() {
+                            return Ok(())
+                        }
+
+                        let i = partial.at(0);
+                        crumb.status = Status::AtChild(i as usize);
+                        self.key_nibbles.push(i);
+
+                        if let Some(child) = nodes[i as usize] {
+                            full_key_nibbles += 1;
+                            partial = partial.mid(1);
+
+                            let prefix = key.back(full_key_nibbles);
+                            self.db.get_raw_or_lookup(child, prefix.left())?
+                        } else {
+                            return Ok(())
+                        }
+                    },
+                    Node::NibbledBranch(slice, nodes, _) => {
+                        if !partial.starts_with(&slice) {
+                            if slice < partial {
+                                crumb.status = Status::Exiting;
+                                self.key_nibbles.append_partial(slice.right());
+                                self.key_nibbles.push((nibble_ops::NIBBLE_LENGTH - 1) as u8);
+                            }
+                            return Ok(());
+                        }
+
+                        full_key_nibbles += slice.len();
+                        partial = partial.mid(slice.len());
+
+                        if partial.is_empty() {
+                            return Ok(())
+                        }
+
+                        let i = partial.at(0);
+                        crumb.status = Status::AtChild(i as usize);
+                        self.key_nibbles.append_partial(slice.right());
+                        self.key_nibbles.push(i);
+
+                        if let Some(child) = nodes[i as usize] {
+                            full_key_nibbles += 1;
+                            partial = partial.mid(1);
+
+                            let prefix = key.back(full_key_nibbles);
+                            self.db.get_raw_or_lookup(child, prefix.left())?
+                        } else {
+                            return Ok(())
+                        }
+                    },
+                    Node::Empty => {
+                        if !partial.is_empty() {
+                            crumb.status = Status::Exiting;
+                        }
+                        return Ok(())
+                    },
+                }
+            };
+
+            node_data = next_node_data;
+            inline = next_inline;
+        }
+    }
+
+    /// Descend into a payload.
+    fn descend<'b, 'c>(&'b mut self, node_data: &'c [u8], inline: bool)
+        -> Result<Node<'c>, TrieHash<L>, CError<L>>
+    {
+        let node_hash = if inline {
+            None
+        } else {
+            Some(L::Hash::hash(node_data))
```

Comment:
The reason for the hash is that it is not trivial to reconstruct the encoded node from an `OwnedNode`, and the hash is valuable. Also, the option indicates whether the node is inline or in the DB.

Reply:
I do not really see the value of the hash, I understand you skip a reencoding this way, but really do not see where it gets useful.
I can trust you on this point, still recalculating the hash here seems like an avoidable calculation.
Maybe having a variant of get_raw_or_lookup that directly returns this Option<Hash> instead of the 'inline' boolean info can do the job (it is the result of try_decode_hash call).